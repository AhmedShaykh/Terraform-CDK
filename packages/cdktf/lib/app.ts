import { Construct, Node, ConstructMetadata } from 'constructs';
import * as fs from 'fs';
import { TerraformStack } from './terraform-stack';
import { version } from '../package.json';
import * as path from 'path';
export const CONTEXT_ENV = 'CDKTF_CONTEXT_JSON';
export interface StackManifest {
  readonly name: string;
  readonly constructPath: string;
  readonly synthesizedStackPath: string;
  readonly workingDirectory: string;
}

export class Manifest {
  public static readonly fileName = 'manifest.json';
  public static readonly stacksFolder = 'stacks';
  public static readonly stackFileName = 'cdk.tf.json';

  public readonly stacks: StackManifest[] = [];

  constructor(public readonly version: string, public readonly outdir: string) {
    fs.mkdirSync(this.outdir, Manifest.stacksFolder)
  }

  public forStack(stack: TerraformStack): StackManifest {
    const node = Node.of(stack)
    const manifest = {
      name: node.id,
      constructPath: node.path,
      workingDirectory: path.join(Manifest.stacksFolder, node.id),
      synthesizedStackPath: path.join(Manifest.stacksFolder, node.id, Manifest.stackFileName)
    }
    this.stacks.push(manifest)

    return manifest;
  }

  public buildManifest(): any {
    return {
      version: this.version,
      stacks: this.stacks.reduce((newObject: Record<string, StackManifest>, stack: StackManifest) => {
        newObject[stack.name] = stack;
        return newObject
      }, {})
    }
  }

  public writeToFile() {
    fs.writeFileSync(path.join(this.outdir, Manifest.fileName), JSON.stringify(this.buildManifest(), undefined, 2));
  }
}

export interface AppOptions {
    /**
     * The directory to output Terraform resources.
     *
     * @default - CDKTF_OUTDIR if defined, otherwise "cdktf.out"
     */
    readonly outdir?: string;
    readonly stackTraces?: boolean;

    /**
     * Additional context values for the application.
     *
     * Context set by the CLI or the `context` key in `cdktf.json` has precedence.
     *
     * Context can be read from any construct using `node.getContext(key)`.
     *
     * @default - no additional context
     */
    readonly context?: { [key: string]: any };
}

/**
 * Represents a cdktf application.
 */
export class App extends Construct {
    /**
     * The output directory into which resources will be synthesized.
     */
    public readonly outdir: string;

    /**
     * The stack which will be synthesized. If not set, all stacks will be synthesized.
     */
    public readonly targetStackId: string | undefined

    /**
     * Defines an app
     * @param options configuration options
     */
    constructor(options: AppOptions = {}) {
        super(undefined as any, '');
        this.outdir = process.env.CDKTF_OUTDIR ?? options.outdir ?? 'cdktf.out';
        this.targetStackId = process.env.CDKTF_TARGET_STACK_ID

        this.loadContext(options.context);

        const node = Node.of(this)
        if (options.stackTraces === false) {
            node.setContext(ConstructMetadata.DISABLE_STACK_TRACE_IN_METADATA, true);
        }

        node.setContext('cdktfVersion', version)
    }

    /**
     * Synthesizes all resources to the output directory
     */
    public synth(): void {
      if (!fs.existsSync(this.outdir)) {
        fs.mkdirSync(this.outdir);
      }

      const manifest = new Manifest(version, this.outdir)

      Node.of(this).synthesize({
        outdir: this.outdir,
        sessionContext: {
          manifest
        }
      });

      manifest.writeToFile();
    }

    private loadContext(defaults: { [key: string]: string } = { }) {
        const node = Node.of(this)

        // prime with defaults passed through constructor
        for (const [k, v] of Object.entries(defaults)) {
          node.setContext(k, v);
        }

        // read from environment
        const contextJson = process.env[CONTEXT_ENV];
        const contextFromEnvironment = contextJson
          ? JSON.parse(contextJson)
          : { };

        for (const [k, v] of Object.entries(contextFromEnvironment)) {
          node.setContext(k, v);
        }
      }
}