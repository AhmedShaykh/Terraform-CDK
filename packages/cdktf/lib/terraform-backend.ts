import { Construct } from "constructs";
import { TerraformElement } from "./terraform-element";

export abstract class TerraformBackend extends TerraformElement {
    constructor(scope: Construct, id: string, protected readonly name: string) {
        super(scope, id);
    }

    // jsii can't handle abstract classes?
    protected synthesizeAttributes(): { [name: string]: any } {
        return {}
    }

    /**
     * Adds this resource to the terraform JSON output.
     */
    public toTerraform(): any {
        return {
            terraform: {
                backend: {
                    [this.name]: { ...this.synthesizeAttributes() }
                }
            }
        };
    }
}