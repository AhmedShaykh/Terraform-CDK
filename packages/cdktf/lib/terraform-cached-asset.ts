import { Construct } from "constructs";
import { App, Manifest } from ".";
import { hashPath } from "./private/fs";
import { TerraformStack } from "./terraform-stack";
import * as path from "path";
import * as fs from "fs";

const CACHED_ASSET_SYMBOL = Symbol.for("cdktf/TerraformCachedAsset");

export interface IBuilder {
  /**
   * TODO: write docstring
   *
   * @throws if the build failed
   */
  build(inputPath: string): void;
}

export interface TerraformCachedAssetConfig {
  builder: IBuilder;
  inputPath: string;
  filePatterns?: string[];
  outputPath: string;
}

/**
 * todo: document
 */
export class TerraformCachedAsset extends Construct {
  constructor(
    scope: Construct,
    id: string,
    private config: TerraformCachedAssetConfig
  ) {
    super(scope, id);
    Object.defineProperty(this, CACHED_ASSET_SYMBOL, { value: true });

    const appOutDir = App.of(this).outdir;
    const terraformFile = path.resolve(
      appOutDir,
      Manifest.stacksFolder,
      TerraformStack.of(this).node.id,
      Manifest.stackFileName
    );

    if (fs.existsSync(terraformFile)) {
      const terraformJson = JSON.parse(fs.readFileSync(terraformFile, "utf-8"));
      const cache = terraformJson["//"].caches;
      if (cache) {
        const hash = cache[TerraformStack.of(this).getLogicalId(this.node)];
        if (hash && hash === this.getHashOfInputDir()) {
          return;
        } else {
          this.config.builder.build(this.config.inputPath);
        }
      } else {
        this.config.builder.build(this.config.inputPath);
      }
    } else {
      this.config.builder.build(this.config.inputPath);
    }
  }

  static isTerraformCachedAsset(x: any): x is TerraformCachedAsset {
    return x !== null && typeof x === "object" && CACHED_ASSET_SYMBOL in x;
  }

  public getHashEntries(): Record<string, string> {
    return {
      [TerraformStack.of(this).getLogicalId(this.node)]:
        this.getHashOfInputDir(),
    };
  }
  private getHashOfInputDir(): string {
    return hashPath(this.config.inputPath, this.config.filePatterns);
  }

  public get hash(): string {
    return this.getHashOfInputDir();
  }
}
