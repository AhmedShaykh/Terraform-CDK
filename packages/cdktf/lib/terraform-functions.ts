// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import { rawString, Token } from ".";
import { FnGenerated } from "./functions/terraform-functions.generated";

// eslint-disable-next-line jsdoc/require-jsdoc
export class Fn extends FnGenerated {
  /**
   * Use this function to wrap a string and escape it properly for the use in Terraform
   * This is only needed in certain scenarios (e.g., if you have unescaped double quotes in the string)
   * @param {String} str
   */
  public static rawString(str: string): string {
    return Token.asString(rawString(str));
  }
}
