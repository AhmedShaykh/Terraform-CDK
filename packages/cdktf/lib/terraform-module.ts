import { Construct } from "constructs";
import { TerraformElement } from "./terraform-element";
import { TerraformProvider } from "./terraform-provider";
import { deepMerge } from "./util";

export interface TerraformModuleOptions {
  readonly source: string;
  readonly version?: string;
  readonly providers?: (TerraformProvider | TerraformModuleProvider)[];
}

export interface TerraformModuleProvider {
  readonly provider: TerraformProvider;
  readonly moduleAlias: string;
}

export abstract class TerraformModule extends TerraformElement {

  public readonly source: string;
  public readonly version?: string;
  private _providers?: (TerraformProvider | TerraformModuleProvider)[];

  constructor(scope: Construct, id: string, options: TerraformModuleOptions) {
    super(scope, id);

    this.source = options.source;
    this.version = options.version;
    this._providers = options.providers;
  }

  // jsii can't handle abstract classes?
  protected synthesizeAttributes(): { [name: string]: any } {
    return {}
  }

  public interpolationForOutput(moduleOutput: string) {
    return `\${module.${this.friendlyUniqueId}.${moduleOutput}}` as any;
  }

  public get providers() {
    return this._providers;
}

public addProvider(provider: TerraformProvider | TerraformModuleProvider) {
    if (!this._providers) {
        this._providers = [];
    }
    this._providers.push(provider);
}

  public toTerraform(): any {
    const attributes = deepMerge({
      ...this.synthesizeAttributes(),
      source: this.source,
      version: this.version,
      providers: this.providers?.map(p => {
        if (p instanceof TerraformProvider) {
            return { [p.terraformResourceType]: p.fqn };
        }
        else {
            return { [`${p.provider.terraformResourceType}.${p.moduleAlias}`]: p.provider.fqn };
        }
    }),
    },
      this.rawOverrides
    )

    attributes['//'] = this.constructNodeMetadata

    return {
      module: {
        [this.friendlyUniqueId]: attributes
      }
    }
  }
}