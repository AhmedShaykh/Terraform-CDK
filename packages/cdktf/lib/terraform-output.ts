import { Construct } from "constructs";
import { TerraformElement } from "./terraform-element";
import { keysToSnakeCase, deepMerge } from "./util";
import { ITerraformDependable } from "./terraform-dependable";
import { Expression, ref } from ".";
import { isArray } from "util";
import { ITerraformAddressable } from "./terraform-addressable";

export interface TerraformOutputConfig {
  readonly value: Expression | ITerraformDependable;
  readonly description?: string;
  readonly sensitive?: boolean;
  readonly dependsOn?: ITerraformDependable[];
  /**
   * If set to true the synthesized Terraform Output will be named after the `id`
   * passed to the constructor instead of the default (TerraformOutput.friendlyUniqueId)
   *
   * @default false
   */
  readonly staticId?: boolean;
}

export class TerraformOutput extends TerraformElement {
  public value: Expression | ITerraformAddressable;
  public description?: string;
  public sensitive?: boolean;
  public dependsOn?: ITerraformDependable[];

  constructor(scope: Construct, id: string, config: TerraformOutputConfig) {
    super(scope, id);

    this.value = config.value;
    this.description = config.description;
    this.sensitive = config.sensitive;
    this.dependsOn = config.dependsOn;
    this.staticId = config.staticId || false;
  }

  public set staticId(staticId: boolean) {
    if (staticId) this.overrideLogicalId(this.node.id);
    else this.resetOverrideLogicalId();
  }

  public get staticId(): boolean {
    return this.friendlyUniqueId === this.node.id;
  }

  private isITerraformAddressable(
    object: any
  ): object is ITerraformAddressable {
    return (
      object &&
      typeof object === "object" &&
      !isArray(object) &&
      "fqn" in object
    );
  }

  protected synthesizeAttributes(): { [key: string]: any } {
    return {
      value: this.isITerraformAddressable(this.value)
        ? ref(this.value.fqn)
        : this.value,
      description: this.description,
      sensitive: this.sensitive,
      dependsOn: this.dependsOn?.map((resource) => `\${${resource.fqn}}`),
    };
  }

  public toTerraform(): any {
    return {
      output: {
        [this.friendlyUniqueId]: deepMerge(
          keysToSnakeCase(this.synthesizeAttributes()),
          this.rawOverrides
        ),
      },
    };
  }
  public toMetadata(): any {
    if (!Object.keys(this.rawOverrides).length) {
      return {};
    }

    return {
      overrides: {
        output: Object.keys(this.rawOverrides),
      },
    };
  }
}
