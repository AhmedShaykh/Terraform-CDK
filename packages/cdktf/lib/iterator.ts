import { TerraformStack } from ".";
import { ComplexList, ComplexMap } from "./complex-computed-list";
import { Fn } from "./terraform-functions";
import { propertyAccess, ref } from "./tfExpression";
import { Lazy, Token } from "./tokens";

export interface IIterator {
  /**
   * @internal used by TerraformResource to set the for_each expression
   */
  getForEachExpression(): any;
}

export abstract class Iterator implements IIterator {
  abstract getForEachExpression(): any;

  /**
   * Creates a new iterator from a list
   */
  public static fromList(
    list: string[] | number[] | ComplexList
  ): ListIterator {
    // TODO: this could return different iterators depending on the type of the list
    // for example it could return a NumberListIterator whose iterator.key would be a number
    return new ListIterator(list);
  }

  /**
   * Creates a new iterator from a map
   */
  public static fromMap(
    map:
      | ComplexMap
      | { [key: string]: any }
      | { [key: string]: string }
      | { [key: string]: number }
      | { [key: string]: boolean }
  ): MapIterator {
    return new MapIterator(map);
  }

  /**
   * @param attribute name of the property to retrieve
   * @returns the given attribute of the current item iterated over as a string
   */
  getString(attribute: string): string {
    return Token.asString(propertyAccess(this.value, [attribute]));
  }

  /**
   * @param attribute name of the property to retrieve
   * @returns the given attribute of the current item iterated over as a number
   */
  getNumber(attribute: string): number {
    return Token.asNumber(propertyAccess(this.value, [attribute]));
  }

  /**
   * @param attribute name of the property to retrieve
   * @returns the given attribute of the current item iterated over as a (string) list
   */
  getList(attribute: string): string[] {
    return Token.asList(propertyAccess(this.value, [attribute]));
  }

  /**
   * @param attribute name of the property to retrieve
   * @returns the given attribute of the current item iterated over as a number list
   */
  getNumberList(attribute: string): number[] {
    return Token.asNumberList(propertyAccess(this.value, [attribute]));
  }

  /**
   * @param attribute name of the property to retrieve
   * @returns the given attribute of the current item iterated over as a map
   */
  getMap(attribute: string): { [key: string]: any } {
    return Token.asAnyMap(propertyAccess(this.value, [attribute]));
  }

  /**
   * @param attribute name of the property to retrieve
   * @returns the given attribute of the current item iterated over as a map of strings
   */
  getStringMap(attribute: string): { [key: string]: string } {
    return Token.asStringMap(propertyAccess(this.value, [attribute]));
  }

  /**
   * @param attribute name of the property to retrieve
   * @returns the given attribute of the current item iterated over as a map of numbers
   */
  getNumberMap(attribute: string): { [key: string]: number } {
    return Token.asNumberMap(propertyAccess(this.value, [attribute]));
  }

  /**
   * @param attribute name of the property to retrieve
   * @returns the given attribute of the current item iterated over as a map of booleans
   */
  getBooleanMap(attribute: string): { [key: string]: boolean } {
    return Token.asBooleanMap(propertyAccess(this.value, [attribute]));
  }

  /**
   * Returns the value of the current item iterated over.
   */
  public get value(): any {
    return Lazy.anyValue(
      {
        produce: (context) =>
          ref("each.value", TerraformStack.of(context.scope)),
      },
      { displayHint: "each.value" }
    );
  }

  public get key(): any {
    return Lazy.anyValue(
      {
        produce: (context) => ref("each.key", TerraformStack.of(context.scope)),
      },
      { displayHint: "each.key" }
    );
  }
}

export class ListIterator extends Iterator {
  constructor(private readonly list: string[] | number[] | ComplexList) {
    super();
  }

  /**
   * Returns the currenty entry in the list or set that is being iterated over.
   * For lists this is the same as `iterator.value`. If you need the index,
   * use count using the escape hatch:
   * https://www.terraform.io/cdktf/concepts/providers-and-resources#escape-hatch
   */
  public get key(): any {
    return super.key;
  }

  /**
   * @internal used by TerraformResource to set the for_each expression
   */
  public getForEachExpression(): any {
    // needs to be wrapped in a set as Terraform only allows sets in for_each
    return Fn.toset(this.list);
  }
}

export class MapIterator extends Iterator {
  constructor(
    private readonly map:
      | ComplexMap
      | { [key: string]: any }
      | { [key: string]: string }
      | { [key: string]: number }
  ) {
    super();
  }

  /**
   * @internal used by TerraformResource to set the for_each expression
   */
  public getForEachExpression(): any {
    // explicit wrapping to circumvent "Found an encoded map token in a scalar string context." error
    return Token.asString(this.map);
  }

  /**
   * Returns the key of the current entry in the map that is being iterated over.
   */
  public get key(): string {
    return super.key;
  }
}
