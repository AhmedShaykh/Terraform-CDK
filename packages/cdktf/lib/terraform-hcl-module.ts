import { Construct } from "constructs";
import { TerraformModuleOptions, TerraformModule } from "./terraform-module";

export interface TerraformHclModuleOptions extends TerraformModuleOptions {
    readonly variables?: {[key: string]: any};
}

export class TerraformHclModule extends TerraformModule {
    private _variables?: {[key: string]: any};

    constructor(scope: Construct, id: string, options: TerraformHclModuleOptions) {
        super(scope, id, options);

        this._variables = options.variables;
    }

    public get variables() {
        return this._variables;
    }

    public set(variable: string, value: any) {
        if (!this._variables) {
            this._variables = {};
        }
        this._variables[variable] = value;
    }

    public get(output: string): any {
        return this.interpolationForOutput(output);
    }

    public getString(output: string): string {
        return this.get(output);
    }

    public getNumber(output: string): number {
        return this.get(output);
    }

    public getBoolean(output: string): boolean {
        return this.get(output);
    }

    public getList(output: string): string[] {
        return this.get(output);
    }

    protected synthesizeAttributes(): { [name: string]: any } {
        return { ...this.variables };
    }
}