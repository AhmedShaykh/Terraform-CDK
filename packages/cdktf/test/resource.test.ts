
import { Testing, TerraformStack } from '../lib';
import { TestProvider, TestResource } from './helper'

test('minimal configuration', () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, 'test');

  new TestResource(stack, 'test', {
    name: 'foo'
  });
  expect(Testing.synth(stack)).toMatchSnapshot();
});

test('with provider alias', () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, 'test');

  new TestProvider(stack, 'provider', {
    accessKey: '1'
  })

  const provider = new TestProvider(stack, 'aliased-provider', {
    alias: "foo",
    accessKey: '123'
  })

  new TestResource(stack, 'test', {
    name: 'bar',
    provider
  });

  expect(Testing.synth(stack)).toMatchSnapshot();
});

test('provider setter', () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, 'test');

  const provider = new TestProvider(stack, 'provider', {
    accessKey: '1'
  })

  const resource = new TestResource(stack, 'test', {
    name: 'bar'
  });

  resource.provider = provider;

  expect(Testing.synth(stack)).toMatchSnapshot();
});
