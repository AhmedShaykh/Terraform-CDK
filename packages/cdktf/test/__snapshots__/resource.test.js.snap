// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`minimal configuration 1`] = `
"{
  \\"resource\\": {
    \\"test_resource\\": {
      \\"testDBA94737\\": {
        \\"name\\": \\"foo\\"
      }
    }
  }
}"
`;

exports[`provider setter 1`] = `
"{
  \\"provider\\": {
    \\"test\\": [
      {
        \\"access_key\\": \\"1\\"
      }
    ]
  },
  \\"resource\\": {
    \\"test_resource\\": {
      \\"testDBA94737\\": {
        \\"name\\": \\"bar\\",
        \\"provider\\": \\"test\\"
      }
    }
  }
}"
`;

exports[`with provider alias 1`] = `
"{
  \\"provider\\": {
    \\"test\\": [
      {
        \\"access_key\\": \\"1\\"
      },
      {
        \\"access_key\\": \\"123\\",
        \\"alias\\": \\"foo\\"
      }
    ]
  },
  \\"resource\\": {
    \\"test_resource\\": {
      \\"testDBA94737\\": {
        \\"name\\": \\"bar\\",
        \\"provider\\": \\"test.foo\\"
      }
    }
  }
}"
`;
